viewing_party/party.py

```diff
@@ -1,23 +1,189 @@
+    
+    for movie_info in friends_movies_list:
+        if movie_info not in user_movies_list:
+            user_hasnt_watched.append(movie_info)
+            
+    user_hasnt_watched_dict = {}
+    for movie in user_hasnt_watched:
+        user_hasnt_watched_dict[movie["title"]] = movie
+    
+    for title, movie in user_hasnt_watched_dict.items():
+        potential_movies.append(movie)
+
+    return potential_movies
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommended_movies_list = []
+#1. loop through the friends movies and see which ones the
+#user hasn't watched.
+    #make a list of user movies
+    user_movies_list = []
+    for movie_info in user_data["watched"]:
+        user_movies_list.append(movie_info)
+    #make a list of friends movies
+    friends_movies_list = []
+    for indi_friends_dict in user_data["friends"]:
+        for key, friend_list in indi_friends_dict.items():
+            for movie_info in friend_list:
+                friends_movies_list.append(movie_info)
+    #if friend movie not in user movie, add to potential watchlist(unique_friend_movies)
+    user_hasnt_watched = [] #list of dictionaries with movie info friends have watched
+    potential_movies = []
+    
+    for movie_info in friends_movies_list:
+        if movie_info not in user_movies_list:
+            user_hasnt_watched.append(movie_info)
+            
+    user_hasnt_watched_dict = {}
+    for movie in user_hasnt_watched:
+        user_hasnt_watched_dict[movie["title"]] = movie
+    
+    for title, movie in user_hasnt_watched_dict.items():
+        potential_movies.append(movie)
+
+#2. loop through the host of the friends unique movies and
+#see which hosts are in the users movies hosts
+    for movie in potential_movies:
+        if movie["host"] in user_data["subscriptions"]:
+            recommended_movies_list.append(movie)
 
+    return recommended_movies_list  
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+
+#Find the most frequently watched genre in users movies
+    freq_genre = get_most_watched_genre(user_data)
+# Find movies the user hasn't watched but at least one of their friend's have watched
+    unique_friends_movies = get_friends_unique_watched(user_data)
+
+# find which movies in the unique_friends_movies list has the same genre as...
+# the users most frequent watched genre
+    rec_movies = []    
+
+    for movie in unique_friends_movies:
+        if freq_genre == None:
+            break
+        if freq_genre in movie["genre"]:
+            rec_movies.append(movie)
+    return rec_movies
+
+def get_rec_from_favorites(user_data):
+    rec_movies = []
+
+    for movie in user_data["favorites"]:
+        if movie in get_unique_watched(user_data):

```

Comment:
It looks like every iteration of your loop calls `get_unique_watched`, which is a lot of work! Instead, you could call `get_unique_watched` once before the for loop and save it in a variable that you can use in the loop instead.